src/uTensor/ops/ActivationFncs_kernels.hpp

```diff
@@ -95,5 +136,119 @@ void relu6_k(Tensor& out, const Tensor& in) {
   }
 }
 
+template <typename T>
+void inplace_softmax_k(Tensor& in, T beta = 1) {
+  T tmp;
+  T mSum = 0;
+  const TensorShape& inShape = in->get_shape();
+  int outer_dim = inShape.num_dims() -1;
+  int depth = inShape[outer_dim];
+  int out_side_numelems = 1;
+  for(int i = 0; i < inShape.num_dims(); i++){
+    out_side_numelems *= (i == outer_dim) ? 1: inShape[i];
+  }
+
+  for (int i = 0; i < out_side_numelems; i++) {
+    // exp(x[i])/sum(exp(x[i])) == exp(x[i]+C)/sum(exp(x[i]+C))
+    T max = std::numeric_limits<T>::lowest();
+    for(int j = 0; j < depth; j++){
+      max = std::max(max, static_cast<T>(in(i, j)));
+    }
+
+    T mSum = 0;
+    for(int j = 0; j < depth; j++){
+      T tmp = exp((static_cast<T>(in(i,j)) - max) * beta);
+      mSum += tmp;
+      in(i,j) = tmp;
+    }
+    for(int j = 0; j < depth; j++){
+      in(i, j)  = static_cast<T>(in(i, j)) / mSum;
+    }
+  }
+}
+template <typename T>
+void softmax_k(Tensor& out, const Tensor& in, T beta=1) {
+  T tmp;
+  T mSum = 0;
+  const TensorShape& inShape = in->get_shape();
+  int outer_dim = inShape.num_dims() -1;
+  int depth = inShape[outer_dim];
+  int out_side_numelems = 1;
+  for(int i = 0; i < inShape.num_dims(); i++){
+    out_side_numelems *= (i == outer_dim) ? 1: inShape[i];
+  }
+
+  for (int i = 0; i < out_side_numelems; i++) {
+    // exp(x[i])/sum(exp(x[i])) == exp(x[i]+C)/sum(exp(x[i]+C))
+    T max = std::numeric_limits<T>::lowest();
+    for(int j = 0; j < depth; j++){
+      max = std::max(max, static_cast<T>(in(i, j)));
+    }
+
+    T mSum = 0;
+    for(int j = 0; j < depth; j++){
+      T tmp = exp((static_cast<T>(in(i,j)) - max) * beta);
+      mSum += tmp;
+      out(i,j) = tmp;
+    }
+    for(int j = 0; j < depth; j++){
+      out(i, j)  = static_cast<T>(out(i, j)) / mSum;
+    }
+  }
+
+}
+
+void sq_softmax_k(Tensor& out, const Tensor& in, int8_t beta=1);
+
+template <typename T>
+class inplace_sigmoid_k {
+  public:
+    void operator()(Tensor& t) const;
+};
+
+template <typename T>
+void inplace_sigmoid_k<T>::operator()(Tensor& t) const {
+  const T one = 1;
+  uint32_t t_size = t->get_shape().get_linear_size();
+  for (uint32_t i = 0; i < t_size; i++) {
+    const T tmp = one / (one + exp(- static_cast<T>(t(i))));
+    t(i) = tmp;
+  }
+}
+
+template <typename T>
+class sigmoid_k_impl {
+  public:
+    void operator()(Tensor& out, const Tensor& in) const;
+
+};
+
+template <typename T>
+void sigmoid_k_impl<T>::operator()(Tensor& out, const Tensor& in) const {
+  const T one = 1;
+  uint32_t t_size = in->get_shape().get_linear_size();
+  for (uint32_t i = 0; i < t_size; i++) {
+    const T tmp = one / (one + exp(- static_cast<T>(in(i))));
+    out(i) = tmp;
+  }
+}
+
+template <>
+void sigmoid_k_impl<int8_t>::operator()(Tensor& out, const Tensor& in) const;
```

Comment:
Just out of curiosity, this will generate symbol for `int8_t` in the static library right?

Reply:
yup yup. This is a specialization of the template for `int_8` types which generates a direct bit of code. That's why it needs a corresponding cpp file :)